### PR TITLE
Utilize libgit2 as git implementation

### PR DESCRIPTION
**Pull request by matthewmcnew**
- All git actions use git2go  
- Updates github actions to build in the golang alpine image with libgit2 installed. 

Closes https://github.com/pivotal/kpack/pull/620
Resolves https://github.com/pivotal/kpack/issues/213

### CODE REVIEW

File: pkg/git/fetch_test.go [hunk 1]
```diff
@@ -18,6 +18,11 @@ import (
 )
 
 func TestGitCheckout(t *testing.T) {
+	testDir, err := ioutil.TempDir("", "test-git2")
```
Comment (matthewmcnew) [inline pkg/git/fetch_test.go:21]: This is not needed

File: pkg/git/git_keychain.go [hunk 2]
```diff
@@ -121,17 +156,36 @@ func NewMountedSecretGitKeychain(volumeName string, basicAuthSecrets, sshAuthSec
...
+		if cred.match(host, allowedTypes) {
+			return cred.git2goCredential(username)
 		}
 	}
-	return anonymousAuth, nil
+	return nil, errors.Errorf("no credentials found for %s", url)
+}
+
+var (
+	isSchemeRegExp   = regexp.MustCompile(`^[^:]+://`)
```
Comment (matthewmcnew) [inline pkg/git/git_keychain.go:176]: Lifted these `regex` from go-git library

File: pkg/git/fetch_test.go [hunk 3]
```diff
@@ -50,18 +49,18 @@ func testGitCheckout(t *testing.T, when spec.G, it spec.S) {
 				err := fetcher.Fetch(testDir, gitUrl, revision, metadataDir)
 				require.NoError(t, err)
 
-				repository, err := gogit.PlainOpenWithOptions(testDir, &gogit.PlainOpenOptions{})
-				require.NoError(t, err)
-
-				worktree, err := repository.Worktree()
+				repository1, err := git2go.InitRepository(testDir, false)
```
Comment (tylerphelan) [inline pkg/git/fetch_test.go:52]: what does 1 represent here?

File: pkg/git/git_keychain.go [hunk 4]
```diff
@@ -1,26 +1,60 @@
...
 	"github.com/pivotal/kpack/pkg/secret"
+	"golang.org/x/crypto/ssh"
 )
 
+type Git2GoCredential interface {
+	Cred() (*git2go.Credential, error)
+}
+
+func asCredentialCallback(gitKeychain GitKeychain) git2go.CredentialsCallback {
+	return func(url string, username_from_url string, allowed_types git2go.CredentialType) (*git2go.Credential, error) {
```
Comment (tylerphelan) [inline pkg/git/git_keychain.go:21]: underscores?

File: pkg/git/fetch_test.go (hunk 3 above)
Comment (matthewmcnew) [reply pkg/git/fetch_test.go:52]: Good Question. Let me fix that.

File: pkg/git/git_keychain.go (hunk 4 above)
Comment (matthewmcnew) [reply pkg/git/git_keychain.go:21]: The underlying git2go implementation uses underscores. Let me switch that over.

File: pkg/git/fetch.go [hunk 5]
```diff
@@ -22,60 +16,55 @@ type Fetcher struct {
...
+		DownloadTags: git2go.DownloadTagsAll,
+		RemoteCallbacks: git2go.RemoteCallbacks{
+			CredentialsCallback: asCredentialCallback(f.Keychain),
+			CertificateCheckCallback: func(cert *git2go.Certificate, valid bool, hostname string) git2go.ErrorCode {
+				return git2go.ErrorCodeOK
+			},
+		},
+	}, "")
+	if err != nil {
+		return errors.Wrap(err, "error fetching remote")
```
Comment (matthewmcnew) [inline pkg/git/fetch.go:46]: error is redundant in `error` 🙂

File: pkg/git/fetch.go [hunk 6]
```diff
@@ -22,60 +16,55 @@ type Fetcher struct {
...
+		return errors.Wrap(err, "initializing repo")
 	}
+	defer repository.Free()
 
-	remote, err := repo.CreateRemote(&config.RemoteConfig{
-		Name: "origin",
-		URLs: []string{gitURL},
+	remote, err := repository.Remotes.CreateWithOptions(gitURL, &git2go.RemoteCreateOptions{
+		Name:  "origin",
+		Flags: git2go.RemoteCreateSkipInsteadof,
```
Comment (matthewmcnew) [inline pkg/git/fetch.go:29]: This is for local to avoid your local git config instead of overwriting the repo url.

File: pkg/git/fetch.go [hunk 7]
```diff
@@ -22,60 +16,55 @@ type Fetcher struct {
...
-	f.Logger.Printf("Cloning %q @ %q...", gitURL, gitRevision)
-	err = remote.Fetch(opts)
-	if err != nil && err != transport.ErrAuthenticationRequired {
-		return errors.Wrap(err, "unable to fetch git repository")
-	} else if err == transport.ErrAuthenticationRequired {
-		return errors.Errorf("invalid credentials to fetch git repository: %s", gitURL)
+	err = remote.Fetch([]string{"refs/*:refs/*"}, &git2go.FetchOptions{
+		DownloadTags: git2go.DownloadTagsAll,
+		RemoteCallbacks: git2go.RemoteCallbacks{
+			CredentialsCallback: asCredentialCallback(f.Keychain),
```
Comment (matthewmcnew) [inline pkg/git/fetch.go:39]: Is the `as` confusing?

File: pkg/git/fetch.go [hunk 8]
```diff
@@ -22,60 +16,55 @@ type Fetcher struct {
...
+			CertificateCheckCallback: func(cert *git2go.Certificate, valid bool, hostname string) git2go.ErrorCode {
+				return git2go.ErrorCodeOK
```
Comment (matthewmcnew) [inline pkg/git/fetch.go:41]: Do we need to validate this cert correctly?

File: pkg/git/remote_git_resolver.go [hunk 9]
```diff
@@ -15,35 +17,63 @@ const defaultRemote = "origin"
...
 }
 
-func (*remoteGitResolver) Resolve(auth transport.AuthMethod, sourceConfig v1alpha1.SourceConfig) (v1alpha1.ResolvedSourceConfig, error) {
-	repo := git.NewRemote(memory.NewStorage(), &config.RemoteConfig{
-		Name: defaultRemote,
-		URLs: []string{sourceConfig.Git.URL},
-	})
-	references, err := repo.List(&git.ListOptions{
-		Auth: auth,
+func (*remoteGitResolver) Resolve(keychain GitKeychain, sourceConfig v1alpha1.SourceConfig) (v1alpha1.ResolvedSourceConfig, error) {
```
Comment (matthewmcnew) [inline pkg/git/remote_git_resolver.go:20]: Let's load test this.

File: pkg/git/fetch.go (hunk 7 above)
Comment (matthewmcnew) [reply pkg/git/fetch.go:39]: @tylerphelan What do you think of `keychainAsCredentialsCallback`?

**Review by tomkennedy513 — APPROVED** [review]
this is awesome work

File: pkg/git/k8s_git_keychain.go [hunk 10]
```diff
@@ -13,22 +12,18 @@ import (
...
-func newK8sGitKeychain(k8sClient k8sclient.Interface) *k8sGitKeychain {
-	return &k8sGitKeychain{secretFetcher: secret.Fetcher{Client: k8sClient}}
+func newK8sGitKeychainFactory(k8sClient k8sclient.Interface) *k8sGitKeychainFactory {
+	return &k8sGitKeychainFactory{secretFetcher: secret.Fetcher{Client: k8sClient}}
 }
 
-func (k *k8sGitKeychain) Resolve(ctx context.Context, namespace, serviceAccount string, git v1alpha1.Git) (transport.AuthMethod, error) {
+func (k *k8sGitKeychainFactory) KeychainForServiceAccount(ctx context.Context, namespace, serviceAccount string) (GitKeychain, error) {
 	secrets, err := k.secretFetcher.SecretsForServiceAccount(ctx, serviceAccount, namespace)
 	if err != nil && !k8serrors.IsNotFound(err) {
```
Comment (tylerphelan) [inline pkg/git/k8s_git_keychain.go:25]: do we care about `IsNotFound` error handling?

**Review by tylerphelan — APPROVED** [review]

File: pkg/git/k8s_git_keychain.go (hunk 10 above)
Comment (matthewmcnew) [reply pkg/git/k8s_git_keychain.go:25]: I believe the idea is a not found error will result in an empty git keychain now.

Comment (matthewmcnew) [reply pkg/git/k8s_git_keychain.go:25]: Previously, this method was the resolve method that would return the actual auth. Now it just returns the keychain so we are delaying the error until libgit2 asks for auth and we can't provide it.

Comment (matthewmcnew) [reply pkg/git/k8s_git_keychain.go:25]: Perhaps we can clarify that though. Let me ponder it.

**Pull request merged**